File: igusto/igusto/page/room_booking/room_booking.js
frappe.pages['room-booking'].on_page_load = function (wrapper) {
    new RoomBooking(wrapper);
};

class RoomBooking {
    constructor(wrapper) {
        this.page = frappe.ui.make_app_page({
            parent: wrapper,
            title: '',
            single_column: true
        });
        this.roomItemsList = [];
        this.serviceItemsList = [];
        this.make();
    }

    load_company_details() {
        frappe.call({
            method: "igusto.igusto.page.room_order.room_order.get_company_details",
            callback: (r) => {
                const data = r.message;
                if (!data) return;

                const logo_html = data.logo
                    ? `<img src="${data.logo}" class="company-logo">`
                    : `<div class="company-logo-placeholder">No Logo</div>`;

                const address_line = "Munnar";
                const contact_line = ` ${data.phone_no || ""} | ${data.email || ""}`;

                const header_html = `
                    <div class="company-header-inner">
                        <div class="company-left">${logo_html}</div>
                        <div class="company-right">
                            <h2 class="company-name">${data.company_name}</h2>
                            <div class="company-details">
                                <div>${address_line} | ${contact_line}</div>
                            </div>
                        </div>
                    </div>
                `;

                $(".company-header").html(header_html);
            }
        });
    }

    make() {
        $(frappe.render_template("room_booking")).appendTo(this.page.main);
        frappe.after_ajax(() => {
            this.load_guest_data();
            this.load_company_details();
            this.load_items();
            this.setup_events();
            this.setup_date_validation();
            // Add default rows after items are loaded
            setTimeout(() => {
                this.addRoomItemRow();
                this.addServiceItemRow();
            }, 500);
        });
    }

    load_guest_data() {
        let storedGuest = localStorage.getItem("guest_data");
        
        if (!storedGuest) {
            frappe.msgprint("⚠️ No guest data found. Please sign up first.");
            setTimeout(() => frappe.set_route("guest-signup"), 1500);
            return;
        }

        try {
            const guestData = JSON.parse(storedGuest);
            console.log("📦 Loaded guest data from localStorage:", guestData);

            frappe.call({
                method: "frappe.client.get_value",
                args: {
                    doctype: "Guest",
                    filters: { name: guestData.guest },
                    fieldname: ["full_name", "nationality"]
                },
                callback: (r) => {
                    if (r.message) {
                        console.log("📦 Fetched from Guest doctype:", r.message);

                        let nationality = r.message.nationality || guestData.nationality || "-";
                        if (nationality.toLowerCase() === 'indian') {
                            nationality = 'India';
                        }
                        
                        const formattedText = `${guestData.guest}: ${r.message.full_name || "-"}, ${nationality}`;
                        $("#guest_full_info").text(formattedText);

                        this.guestData = {
                            ...guestData,
                            full_name: r.message.full_name,
                            nationality: r.message.nationality || guestData.nationality
                        };

                        console.log("✅ Guest data loaded successfully");
                    } else {
                        frappe.msgprint("⚠️ Guest not found in system.");
                        setTimeout(() => frappe.set_route("guest-signup"), 1500);
                    }
                }
            });

        } catch (e) {
            console.error("❌ Failed to parse guest data:", e);
            frappe.msgprint("⚠️ Invalid guest data. Please sign up again.");
            setTimeout(() => frappe.set_route("guest-signup"), 1500);
        }
    }

    load_items() {
        // Load Room items (Item Group = "Room" or parent = "Room")
        frappe.call({
            method: "frappe.client.get_list",
            args: {
                doctype: "Item",
                filters: [
                    ["is_sales_item", "=", 1],
                    ["item_group", "in", ["Room", "Rooms"]]
                ],
                fields: ["name", "item_name", "stock_uom", "standard_rate", "item_group"]
            },
            callback: (r) => {
                if (r.message) {
                    this.roomItemsList = r.message;
                    console.log("✅ Room Items loaded:", this.roomItemsList.length);
                }
            }
        });

        // Load Other Service items (excluding Room items)
        frappe.call({
            method: "frappe.client.get_list",
            args: {
                doctype: "Item",
                filters: [
                    ["is_sales_item", "=", 1],
                    ["item_group", "not in", ["Room", "Rooms"]]
                ],
                fields: ["name", "item_name", "stock_uom", "standard_rate", "item_group"]
            },
            callback: (r) => {
                if (r.message) {
                    this.serviceItemsList = r.message;
                    console.log("✅ Service Items loaded:", this.serviceItemsList.length);
                }
            }
        });
    }

    setup_date_validation() {
        const $form = $(".booking-card");
        const today = frappe.datetime.get_today();
        let me = this;

        $form.find("#check_in").attr("min", today);
        
        $form.find("#check_in").on("change", function() {
            const checkIn = $(this).val();
            
            if (checkIn && checkIn < today) {
                frappe.msgprint("⚠️ Start date cannot be in the past.");
                $(this).val("");
                return;
            }

            if (checkIn) {
                $form.find("#check_out").attr("min", checkIn);
                
                const checkOut = $form.find("#check_out").val();
                if (checkOut && checkOut < checkIn) {
                    $form.find("#check_out").val("");
                }
            }

            me.calculateTotalDays();
        });

        $form.find("#check_out").on("change", function() {
            const checkOut = $(this).val();
            const checkIn = $form.find("#check_in").val();

            if (checkOut && checkOut < today) {
                frappe.msgprint("⚠️ End date cannot be in the past.");
                $(this).val("");
                return;
            }

            if (checkIn && checkOut && checkOut < checkIn) {
                frappe.msgprint("⚠️ End date must be same or after start date.");
                $(this).val("");
                return;
            }

            me.calculateTotalDays();
        });

        // Total Rooms change event
        $form.find("#total_rooms").on("input", function() {
            me.updateAllItemsQty();
        });
    }

    calculateTotalDays() {
        const checkIn = $("#check_in").val();
        const checkOut = $("#check_out").val();

        if (!checkIn || !checkOut) {
            $("#total_days").val("");
            return;
        }

        const start = new Date(checkIn);
        const end = new Date(checkOut);
        const diffTime = Math.abs(end - start);
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
        const totalDays = diffDays || 1;

        $("#total_days").val(totalDays);
        
        // Update qty for all items
        this.updateAllItemsQty();
    }

    updateAllItemsQty() {
        const totalRooms = parseInt($("#total_rooms").val()) || 1;
        const totalDays = parseInt($("#total_days").val()) || 0;

        if (totalDays === 0) return;

        const calculatedQty = totalRooms * totalDays;

        // Update Room Items
        $("#room_items_body tr").each(function() {
            const $row = $(this);
            const $qtyInput = $row.find(".item-qty");
            $qtyInput.val(calculatedQty);
            
            const rate = parseFloat($row.find(".item-rate").val()) || 0;
            const amount = calculatedQty * rate;
            $row.find(".item-amount").val(amount.toFixed(2));
        });

        // Update Service Items
        $("#service_items_body tr").each(function() {
            const $row = $(this);
            const $qtyInput = $row.find(".item-qty");
            $qtyInput.val(calculatedQty);
            
            const rate = parseFloat($row.find(".item-rate").val()) || 0;
            const amount = calculatedQty * rate;
            $row.find(".item-amount").val(amount.toFixed(2));
        });
    }

    setup_events() {
        const $form = $(".booking-card");
        let me = this;

        // Add Room Item button
        $form.on("click", "#add_room_item", function(e) {
            e.preventDefault();
            me.addRoomItemRow();
        });

        // Add Service Item button
        $form.on("click", "#add_service_item", function(e) {
            e.preventDefault();
            me.addServiceItemRow();
        });

        // Room Item selection change
        $form.on("change", ".room-item-select", function() {
            const $row = $(this).closest("tr");
            const itemCode = $(this).val();
            
            if (!itemCode) return;

            const item = me.roomItemsList.find(i => i.name === itemCode);
            if (item) {
                $row.find(".item-uom").val(item.stock_uom || "");
                $row.find(".item-rate").val(item.standard_rate || 0);

                const totalRooms = parseInt($("#total_rooms").val()) || 1;
                const totalDays = parseInt($("#total_days").val()) || 0;
                
                if (totalDays > 0) {
                    const qty = totalRooms * totalDays;
                    $row.find(".item-qty").val(qty);
                    
                    const amount = qty * (item.standard_rate || 0);
                    $row.find(".item-amount").val(amount.toFixed(2));
                } else {
                    $row.find(".item-qty").val(1);
                    $row.find(".item-amount").val((item.standard_rate || 0).toFixed(2));
                }
            }
        });

        // Service Item selection change
        $form.on("change", ".service-item-select", function() {
            const $row = $(this).closest("tr");
            const itemCode = $(this).val();
            
            if (!itemCode) return;

            const item = me.serviceItemsList.find(i => i.name === itemCode);
            if (item) {
                $row.find(".item-uom").val(item.stock_uom || "");
                $row.find(".item-rate").val(item.standard_rate || 0);

                const totalRooms = parseInt($("#total_rooms").val()) || 1;
                const totalDays = parseInt($("#total_days").val()) || 0;
                
                if (totalDays > 0) {
                    const qty = totalRooms * totalDays;
                    $row.find(".item-qty").val(qty);
                    
                    const amount = qty * (item.standard_rate || 0);
                    $row.find(".item-amount").val(amount.toFixed(2));
                } else {
                    $row.find(".item-qty").val(1);
                    $row.find(".item-amount").val((item.standard_rate || 0).toFixed(2));
                }
            }
        });

        // Qty change for both tables
        $form.on("input", ".item-qty", function() {
            const $row = $(this).closest("tr");
            const qty = parseFloat($(this).val()) || 0;
            const rate = parseFloat($row.find(".item-rate").val()) || 0;
            const amount = qty * rate;
            $row.find(".item-amount").val(amount.toFixed(2));
        });

        // Remove item from Room Items
        $form.on("click", "#room_items_body .remove-item-btn", function() {
            $(this).closest("tr").remove();
            me.reindexItems("#room_items_body");
        });

        // Remove item from Service Items
        $form.on("click", "#service_items_body .remove-item-btn", function() {
            $(this).closest("tr").remove();
            me.reindexItems("#service_items_body");
        });

        // Book Now Button
        $form.on("click", "#create_booking_btn", function (e) {
            e.preventDefault();

            if (!me.guestData) {
                frappe.msgprint("⚠️ Guest data not loaded. Please go back and sign up.");
                return;
            }

            const bookingData = {
                guest: me.guestData.guest,
                mobile: me.guestData.mobile,
                email: me.guestData.email,
                nationality: me.guestData.nationality,
                check_in: $form.find("#check_in").val() || "",
                check_out: $form.find("#check_out").val() || "",
                total_adults: $form.find("#total_adults_input").val() || "1",
                total_children: $form.find("#total_children_input").val() || "0",
                total_rooms: $form.find("#total_rooms").val() || "1",
                total_days: $form.find("#total_days").val() || "0"
            };

            // Validation
            if (!bookingData.check_in || !bookingData.check_out) {
                frappe.msgprint("⚠️ Start date and End date are required.");
                return;
            }
            if (!bookingData.total_adults || parseInt(bookingData.total_adults) < 1) {
                frappe.msgprint("⚠️ Total adults must be at least 1.");
                $form.find("#total_adults_input").focus();
                return;
            }
            if (!bookingData.total_rooms || parseInt(bookingData.total_rooms) < 1) {
                frappe.msgprint("⚠️ Total rooms must be at least 1.");
                $form.find("#total_rooms").focus();
                return;
            }

            // Collect Room Items
            const roomItems = [];
            $("#room_items_body tr").each(function() {
                const $row = $(this);
                const itemCode = $row.find(".room-item-select").val();
                
                if (itemCode) {
                    roomItems.push({
                        item_code: itemCode,
                        qty: parseFloat($row.find(".item-qty").val()) || 0,
                        uom: $row.find(".item-uom").val() || "",
                        rate: parseFloat($row.find(".item-rate").val()) || 0,
                        amount: parseFloat($row.find(".item-amount").val()) || 0
                    });
                }
            });

            // Collect Service Items
            const serviceItems = [];
            $("#service_items_body tr").each(function() {
                const $row = $(this);
                const itemCode = $row.find(".service-item-select").val();
                
                if (itemCode) {
                    serviceItems.push({
                        item_code: itemCode,
                        qty: parseFloat($row.find(".item-qty").val()) || 0,
                        uom: $row.find(".item-uom").val() || "",
                        rate: parseFloat($row.find(".item-rate").val()) || 0,
                        amount: parseFloat($row.find(".item-amount").val()) || 0
                    });
                }
            });

            if (roomItems.length === 0) {
                frappe.msgprint("⚠️ Please add at least one room item.");
                return;
            }

            const no_of_guests = parseInt(bookingData.total_adults) + parseInt(bookingData.total_children);

            console.log("📤 Sending booking data:", {
                ...bookingData,
                no_of_guests,
                room_items: roomItems,
                service_items: serviceItems
            });

            frappe.call({
                method: "igusto.igusto.page.room_booking.room_booking.create_booking",
                args: {
                    guest: bookingData.guest,
                    mobile: bookingData.mobile,
                    email: bookingData.email,
                    nationality: bookingData.nationality,
                    check_in: bookingData.check_in,
                    check_out: bookingData.check_out,
                    no_of_guests: no_of_guests,
                    total_adults: bookingData.total_adults,
                    total_children: bookingData.total_children,
                    total_rooms: bookingData.total_rooms,
                    total_days: bookingData.total_days,
                    room_items: JSON.stringify(roomItems),
                    service_items: JSON.stringify(serviceItems)
                },
                callback: function (r) {
                    if (r.message) {
                        frappe.msgprint({
                            title: "✅ Success",
                            message: `Room Booking Reserved Successfully!<br>Sales Order: ${r.message.sales_order}`,
                            indicator: "green"
                        });
                        
                        localStorage.removeItem("guest_data");
                        console.log("🗑️ Cleared guest_data from localStorage");
                    }
                },
                error: function (err) {
                    console.error("❌ Booking Error:", err);
                    frappe.msgprint("❌ Failed to create booking.");
                }
            });
        });

        // Cancel button
        $form.on("click", "#back_btn", function () {
            frappe.set_route("guest-signup");
        });
    }

    addRoomItemRow() {
        const rowIndex = $("#room_items_body tr").length + 1;
        
        let itemOptions = '<option value="">Select Room Item</option>';
        this.roomItemsList.forEach(item => {
            itemOptions += `<option value="${item.name}">${item.item_name || item.name}</option>`;
        });

        const row = `
            <tr>
                <td class="sr-no">${rowIndex}</td>
                <td>
                    <select class="service-item-select room-item-select">
                        ${itemOptions}
                    </select>
                </td>
                <td>
                    <input type="number" class="service-item-input item-qty" value="1" min="0" step="0.01">
                </td>
                <td>
                    <input type="text" class="service-item-input item-uom" readonly>
                </td>
                <td>
                    <input type="number" class="service-item-input item-rate" value="0" min="0" step="0.01" readonly>
                </td>
                <td>
                    <input type="number" class="service-item-input item-amount" value="0" readonly>
                </td>
                <td>
                    <button type="button" class="remove-item-btn">×</button>
                </td>
            </tr>
        `;

        $("#room_items_body").append(row);
    }

    addServiceItemRow() {
        const rowIndex = $("#service_items_body tr").length + 1;
        
        let itemOptions = '<option value="">Select Service Item</option>';
        this.serviceItemsList.forEach(item => {
            itemOptions += `<option value="${item.name}">${item.item_name || item.name}</option>`;
        });

        const row = `
            <tr>
                <td class="sr-no">${rowIndex}</td>
                <td>
                    <select class="service-item-select">
                        ${itemOptions}
                    </select>
                </td>
                <td>
                    <input type="number" class="service-item-input item-qty" value="1" min="0" step="0.01">
                </td>
                <td>
                    <input type="text" class="service-item-input item-uom" readonly>
                </td>
                <td>
                    <input type="number" class="service-item-input item-rate" value="0" min="0" step="0.01" readonly>
                </td>
                <td>
                    <input type="number" class="service-item-input item-amount" value="0" readonly>
                </td>
                <td>
                    <button type="button" class="remove-item-btn">×</button>
                </td>
            </tr>
        `;

        $("#service_items_body").append(row);
    }

    reindexItems(tableBodySelector) {
        $(tableBodySelector + " tr").each(function(index) {
            $(this).find(".sr-no").text(index + 1);
        });
    }
}

File: igusto/igusto/page/room_booking/room_booking.py
import frappe
from frappe import _
from datetime import datetime
import json

@frappe.whitelist()
def create_booking(guest, mobile, email, nationality, check_in, check_out, 
                   no_of_guests, total_adults, total_children, total_rooms, 
                   total_days, room_items=None, service_items=None):
    """Create Sales Order for room booking with room items and service items"""
    
    try:
        # ------------------------------- 
        # Validation: Check dates
        # ------------------------------- 
        validate_dates(check_in, check_out)
        
        # ------------------------------- 
        # Validate guest counts
        # ------------------------------- 
        validate_guest_counts(total_adults, total_children, no_of_guests)
        
        # ------------------------------- 
        # Validate rooms and days
        # ------------------------------- 
        validate_rooms_and_days(total_rooms, total_days)
        
        # ------------------------------- 
        # Fetch Guest
        # ------------------------------- 
        if not frappe.db.exists("Guest", guest):
            frappe.throw(_("Guest {0} does not exist").format(guest))
        
        guest_doc = frappe.get_doc("Guest", guest)
        full_name = guest_doc.full_name
        guest_mobile = guest_doc.primary_contact or mobile
        guest_email = guest_doc.email or email
        guest_nationality = guest_doc.nationality or nationality
        
        # ------------------------------- 
        # Create Customer if not exists
        # ------------------------------- 
        customer = get_or_create_customer(full_name, guest_mobile, guest_email)
        
        # ------------------------------- 
        # Determine currency
        # ------------------------------- 
        final_currency = "INR" if guest_nationality and guest_nationality.lower() == "india" else "USD"
        
        # ------------------------------- 
        # Prepare items list
        # ------------------------------- 
        items_list = []
        
        # ------------------------------- 
        # Add Room Items
        # ------------------------------- 
        if room_items:
            try:
                room_items_list = json.loads(room_items) if isinstance(room_items, str) else room_items
                
                for room_item in room_items_list:
                    item_code = room_item.get("item_code")
                    qty = float(room_item.get("qty", 0))
                    rate = float(room_item.get("rate", 0))
                    uom = room_item.get("uom", "")
                    
                    if item_code and qty > 0:
                        # Verify item exists
                        if frappe.db.exists("Item", item_code):
                            items_list.append({
                                "item_code": item_code,
                                "qty": qty,
                                "rate": rate,
                                "uom": uom,
                                "description": f"Room Item for {full_name}\n"
                                               f"Check-in: {check_in} | Check-out: {check_out}\n"
                                               f"Total Rooms: {total_rooms} | Total Days: {total_days}\n"
                                               f"Adults: {total_adults} | Children: {total_children}"
                            })
                        else:
                            frappe.log_error(f"Room Item {item_code} not found", "Room Item Missing")
                            
            except Exception as e:
                frappe.log_error(f"Error parsing room items: {str(e)}", "Room Items Parse Error")
        
        # Validation: At least one room item is required
        if not items_list:
            frappe.throw(_("At least one room item is required"))
        
        # ------------------------------- 
        # Add Service Items
        # ------------------------------- 
        if service_items:
            try:
                service_items_list = json.loads(service_items) if isinstance(service_items, str) else service_items
                
                for service_item in service_items_list:
                    item_code = service_item.get("item_code")
                    qty = float(service_item.get("qty", 0))
                    rate = float(service_item.get("rate", 0))
                    uom = service_item.get("uom", "")
                    
                    if item_code and qty > 0:
                        # Verify item exists
                        if frappe.db.exists("Item", item_code):
                            items_list.append({
                                "item_code": item_code,
                                "qty": qty,
                                "rate": rate,
                                "uom": uom,
                                "description": f"Service Item for {full_name}\n"
                                               f"Total Rooms: {total_rooms} | Total Days: {total_days}"
                            })
                        else:
                            frappe.log_error(f"Service Item {item_code} not found", "Service Item Missing")
                            
            except Exception as e:
                frappe.log_error(f"Error parsing service items: {str(e)}", "Service Items Parse Error")
        
        # ------------------------------- 
        # Create Sales Order
        # ------------------------------- 
        so = frappe.get_doc({
            "doctype": "Sales Order",
            "customer": customer.name,
            "custom_number_of_guests": no_of_guests,
            "custom_guest_id": guest,
            "order_type": "Service",
            "transaction_date": check_in,
            "delivery_date": check_out,
            "currency": final_currency,
            "items": items_list
        })
        
        so.insert(ignore_permissions=True)
        frappe.db.commit()
        
        frappe.logger().info(f"✅ Sales Order {so.name} created for Guest: {full_name}")
        
        # ------------------------------- 
        # Return Sales Order details
        # ------------------------------- 
        return {
            "sales_order": so.name,
            "message": "Booking reserved successfully",
            "customer": customer.name,
            "total_amount": so.grand_total,
            "items_count": len(items_list),
            "total_rooms": total_rooms,
            "total_days": total_days
        }
        
    except Exception as e:
        frappe.log_error(f"Booking Error: {str(e)}", "Room Booking")
        frappe.throw(_("Failed to create booking: {0}").format(str(e)))


def validate_dates(check_in, check_out):
    """Validate check-in and check-out dates"""
    
    if not check_in or not check_out:
        frappe.throw(_("Check-in and Check-out dates are required"))
    
    try:
        check_in_date = datetime.strptime(check_in, "%Y-%m-%d").date()
        check_out_date = datetime.strptime(check_out, "%Y-%m-%d").date()
        today = datetime.now().date()
        
        # Check if check-in is in the past
        if check_in_date < today:
            frappe.throw(_("Check-in date cannot be in the past"))
        
        # Check if check-out is in the past
        if check_out_date < today:
            frappe.throw(_("Check-out date cannot be in the past"))
        
        # Check if check-out is before check-in (equal is allowed)
        if check_out_date < check_in_date:
            frappe.throw(_("Check-out date must be same or after check-in date"))
            
    except ValueError:
        frappe.throw(_("Invalid date format. Use YYYY-MM-DD"))


def validate_guest_counts(total_adults, total_children, no_of_guests):
    """Validate guest count logic"""
    
    try:
        adults = int(total_adults)
        children = int(total_children)
        total = int(no_of_guests)
        
        if adults < 1:
            frappe.throw(_("At least 1 adult is required"))
        
        if children < 0:
            frappe.throw(_("Total children cannot be negative"))
        
        if (adults + children) != total:
            frappe.throw(_("Total guests mismatch. Adults + Children should equal Total Guests"))
            
    except ValueError:
        frappe.throw(_("Invalid guest count values"))


def validate_rooms_and_days(total_rooms, total_days):
    """Validate total rooms and total days"""
    
    try:
        rooms = int(total_rooms)
        days = int(total_days)
        
        if rooms < 1:
            frappe.throw(_("Total rooms must be at least 1"))
        
        if days < 1:
            frappe.throw(_("Total days must be at least 1"))
            
    except ValueError:
        frappe.throw(_("Invalid rooms or days values"))


def get_or_create_customer(full_name, mobile, email):
    """Get existing customer or create new one"""
    
    customer_name = full_name
    
    if frappe.db.exists("Customer", {"customer_name": customer_name}):
        customer = frappe.get_doc("Customer", {"customer_name": customer_name})
    else:
        customer = frappe.get_doc({
            "doctype": "Customer",
            "customer_name": customer_name,
            "customer_type": "Individual",
            "customer_group": "Individual",
            "territory": "India",
            "mobile_no": mobile,
            "email_id": email
        })
        customer.insert(ignore_permissions=True)
        frappe.db.commit()
        frappe.logger().info(f"✅ Customer created: {customer_name}")
    
    return customer


@frappe.whitelist()
def get_room_items():
    """Get list of room items (Item Group = Room or parent = Room)"""
    
    room_items = frappe.get_all(
        "Item",
        filters={
            "is_sales_item": 1,
            "item_group": ["in", ["Room", "Rooms"]]
        },
        fields=["name", "item_name", "stock_uom", "standard_rate", "item_group"],
        order_by="item_name"
    )
    
    return room_items


@frappe.whitelist()
def get_service_items():
    """Get list of service items (excluding Room items)"""
    
    service_items = frappe.get_all(
        "Item",
        filters={
            "is_sales_item": 1,
            "item_group": ["not in", ["Room", "Rooms"]]
        },
        fields=["name", "item_name", "stock_uom", "standard_rate", "item_group"],
        order_by="item_name"
    )
    
    return service_items